File: app/src/main/java/cl/ufchile/app/domain/engine/MortgageEngine.kt
package cl.ufchile.app.domain.engine

import cl.ufchile.app.domain.model.AmortizationRow
import cl.ufchile.app.domain.model.MortgageInput
import cl.ufchile.app.domain.model.MortgageResult
import cl.ufchile.app.domain.model.PrepaymentMode
import cl.ufchile.app.domain.model.RateConvention
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

/**
 * Builds the amortisation schedule of a UF-denominated mortgage (UF + x%).
 *
 * Uses the French system: a constant capital-plus-interest payment. Insurance
 * is charged on top of that payment and is therefore *not* constant, since the
 * desgravamen premium follows the outstanding balance.
 *
 * All money is BigDecimal in UF at [MONEY_SCALE] decimals. Nothing here uses
 * Double for money; doubles appear only where a rate is being solved for.
 */
object MortgageEngine {

    /** UF amounts are quoted with four decimals in Chilean payment tables. */
    const val MONEY_SCALE = 4
    private const val RATE_SCALE = 12
    private val MC = MathContext.DECIMAL64
    private val HUNDRED = BigDecimal(100)

    fun monthlyRate(annualRatePct: BigDecimal, convention: RateConvention): BigDecimal {
        val annual = annualRatePct.divide(HUNDRED, RATE_SCALE, RoundingMode.HALF_UP)
        return when (convention) {
            RateConvention.NOMINAL_DIVIDED ->
                annual.divide(BigDecimal(12), RATE_SCALE, RoundingMode.HALF_UP)

            RateConvention.EFFECTIVE_EQUIVALENT -> {
                // The twelfth root has no exact BigDecimal form; solving it in
                // double is fine because this is a rate, not an amount.
                val m = Math.pow(1.0 + annual.toDouble(), 1.0 / 12.0) - 1.0
                BigDecimal(m).setScale(RATE_SCALE, RoundingMode.HALF_UP)
            }
        }
    }

    /** Constant French payment for [principal] at [i] per month over [n] months. */
    fun payment(principal: BigDecimal, i: BigDecimal, n: Int): BigDecimal {
        if (n <= 0) return BigDecimal.ZERO.setScale(MONEY_SCALE)
        if (i.signum() == 0) {
            return principal.divide(BigDecimal(n), MONEY_SCALE, RoundingMode.HALF_UP)
        }
        val onePlusI = BigDecimal.ONE + i
        val discount = BigDecimal.ONE.divide(onePlusI.pow(n, MC), MC) // (1+i)^-n
        return principal.multiply(i)
            .divide(BigDecimal.ONE - discount, MONEY_SCALE, RoundingMode.HALF_UP)
    }

    fun simulate(input: MortgageInput): MortgageResult {
        val i = monthlyRate(input.annualRatePct, input.rateConvention)
        val principal = input.loanAmountUf.setScale(MONEY_SCALE, RoundingMode.HALF_UP)
        val n = input.termMonths
        val basePayment = payment(principal, i, n)

        val lifeRate = input.lifeInsuranceMonthlyPct.divide(HUNDRED, RATE_SCALE, RoundingMode.HALF_UP)
        val fire = input.fireInsuranceMonthlyUf.setScale(MONEY_SCALE, RoundingMode.HALF_UP)
        val prepaysByMonth = input.prepayments.groupBy { it.monthNumber }

        val rows = ArrayList<AmortizationRow>(n)
        var balance = principal
        var currentPayment = basePayment
        var month = 1
        // A prepayment can only ever shorten the schedule, so n is a hard cap.
        while (balance.signum() > 0 && month <= n) {
            val opening = balance
            var interest = opening.multiply(i).setScale(MONEY_SCALE, RoundingMode.HALF_UP)
            var principalPart = (currentPayment - interest).setScale(MONEY_SCALE, RoundingMode.HALF_UP)
            var thisPayment = currentPayment

            // Final instalment: settle whatever is left rather than overshoot.
            if (principalPart >= opening) {
                principalPart = opening
                thisPayment = (principalPart + interest).setScale(MONEY_SCALE, RoundingMode.HALF_UP)
            }
            // A negative-amortising payment would never repay the loan.
            if (principalPart.signum() < 0) principalPart = BigDecimal.ZERO.setScale(MONEY_SCALE)

            val life = opening.multiply(lifeRate).setScale(MONEY_SCALE, RoundingMode.HALF_UP)
            var afterPrincipal = (opening - principalPart).setScale(MONEY_SCALE, RoundingMode.HALF_UP)

            var prepaid = BigDecimal.ZERO.setScale(MONEY_SCALE)
            val prepays = prepaysByMonth[month]
            if (prepays != null && afterPrincipal.signum() > 0) {
                for (p in prepays) {
                    val applied = p.amountUf.setScale(MONEY_SCALE, RoundingMode.HALF_UP)
                        .min(afterPrincipal)
                    if (applied.signum() <= 0) continue
                    prepaid = prepaid + applied
                    afterPrincipal = afterPrincipal - applied
                    if (p.mode == PrepaymentMode.REDUCE_PAYMENT) {
                        val remaining = n - month
                        currentPayment =
                            if (remaining > 0) payment(afterPrincipal, i, remaining)
                            else afterPrincipal
                    }
                    // REDUCE_TERM keeps the payment; the loop simply ends sooner.
                }
            }

            val closing = afterPrincipal
            val outflow = (thisPayment + life + fire + prepaid)
                .setScale(MONEY_SCALE, RoundingMode.HALF_UP)

            rows.add(
                AmortizationRow(
                    number = month,
                    date = input.startDate.plusMonths(month.toLong()),
                    openingBalanceUf = opening,
                    interestUf = interest,
                    principalUf = principalPart,
                    paymentUf = thisPayment,
                    lifeInsuranceUf = life,
                    fireInsuranceUf = fire,
                    prepaymentUf = prepaid,
                    totalOutflowUf = outflow,
                    closingBalanceUf = closing,
                )
            )
            balance = closing
            month++
        }

        val stampTax = principal.multiply(input.stampTaxPct)
            .divide(HUNDRED, MONEY_SCALE, RoundingMode.HALF_UP)
        val upfront = (input.originationFeeUf + stampTax + input.otherUpfrontCostsUf)
            .setScale(MONEY_SCALE, RoundingMode.HALF_UP)

        val totalInterest = rows.sumUf { it.interestUf }
        val totalLife = rows.sumUf { it.lifeInsuranceUf }
        val totalFire = rows.sumUf { it.fireInsuranceUf }
        val totalPrepay = rows.sumUf { it.prepaymentUf }
        val totalOutflow = rows.sumUf { it.totalOutflowUf }

        return MortgageResult(
            input = input,
            loanAmountUf = principal,
            monthlyRate = i,
            basePaymentUf = basePayment,
            firstTotalPaymentUf = rows.firstOrNull()?.totalOutflowUf
                ?: BigDecimal.ZERO.setScale(MONEY_SCALE),
            schedule = rows,
            totalInterestUf = totalInterest,
            totalLifeInsuranceUf = totalLife,
            totalFireInsuranceUf = totalFire,
            totalPrepaymentsUf = totalPrepay,
            upfrontCostsUf = upfront,
            totalCostUf = (totalOutflow + upfront).setScale(MONEY_SCALE, RoundingMode.HALF_UP),
            caePct = solveCae(principal, upfront, rows.map { it.totalOutflowUf }),
            effectiveTermMonths = rows.size,
            monthsSaved = (n - rows.size).coerceAtLeast(0),
        )
    }

    /**
     * Carga Anual Equivalente: the annual rate that makes the present value of
     * everything the borrower pays equal the net amount actually received.
     *
     * Solved by bisection on the monthly rate. Returns null when the cash flows
     * have no sign change and therefore no root.
     */
    fun solveCae(
        principal: BigDecimal,
        upfrontCosts: BigDecimal,
        outflows: List<BigDecimal>,
    ): BigDecimal? {
        if (outflows.isEmpty() || principal.signum() <= 0) return null
        val net = (principal - upfrontCosts).toDouble()
        if (net <= 0.0) return null
        val flows = outflows.map { it.toDouble() }

        fun npv(monthly: Double): Double {
            var acc = net
            var discount = 1.0
            val step = 1.0 + monthly
            for (f in flows) {
                discount *= step
                acc -= f / discount
            }
            return acc
        }

        var lo = -0.9999
        var hi = 5.0
        if (npv(lo) * npv(hi) > 0) return null
        repeat(200) {
            val mid = (lo + hi) / 2
            if (npv(lo) * npv(mid) <= 0) hi = mid else lo = mid
        }
        val monthly = (lo + hi) / 2
        val annual = (Math.pow(1.0 + monthly, 12.0) - 1.0) * 100.0
        if (annual.isNaN() || annual.isInfinite()) return null
        return BigDecimal(annual).setScale(2, RoundingMode.HALF_UP)
    }

    private inline fun List<AmortizationRow>.sumUf(sel: (AmortizationRow) -> BigDecimal): BigDecimal =
        fold(BigDecimal.ZERO) { acc, r -> acc + sel(r) }
            .setScale(MONEY_SCALE, RoundingMode.HALF_UP)
}

File: web/src/domain/mortgageEngine.ts
import {
  HALF_UP, HUNDRED, ONE, ZERO, divScale, minOf, money, scale, sumScaled, type Money,
} from "./money.ts";
import { plusMonths } from "./dates.ts";
import {
  loanAmountUf, termMonths,
  type AmortizationRow, type MortgageInput, type MortgageResult, type Prepayment,
  type RateConventionKey,
} from "./models.ts";

/**
 * Builds the amortisation schedule of a UF-denominated mortgage (UF + x%).
 *
 * French system: a constant capital-plus-interest payment. Insurance is charged
 * on top and is deliberately not constant, since the desgravamen premium
 * follows the outstanding balance.
 *
 * Every amount is a decimal at MONEY_SCALE places. Nothing here uses a float
 * for money; floats appear only where a rate is being solved for.
 */
export const MONEY_SCALE = 4;
const RATE_SCALE = 12;
/** (1+i)^n keeps about this many places, matching the Kotlin's DECIMAL64. */
const POWER_SCALE = 20;

export const monthlyRate = (annualRatePct: Money, convention: RateConventionKey): Money => {
  const annual = divScale(annualRatePct, HUNDRED, RATE_SCALE);
  if (convention === "NOMINAL_DIVIDED") return divScale(annual, money(12), RATE_SCALE);
  // The twelfth root has no exact decimal form; solving it in a float is fine
  // because this is a rate, not an amount.
  const m = Math.pow(1 + Number(annual.toString()), 1 / 12) - 1;
  return scale(money(m.toFixed(RATE_SCALE)), RATE_SCALE);
};

/** Constant French payment for [principal] at [i] per month over [n] months. */
export const payment = (principal: Money, i: Money, n: number): Money => {
  if (n <= 0) return scale(ZERO, MONEY_SCALE);
  if (i.cmp(0) === 0) return divScale(principal, money(n), MONEY_SCALE);
  const powed = ONE.plus(i).pow(n).round(POWER_SCALE, HALF_UP);
  const discount = ONE.div(powed); // (1+i)^-n
  return divScale(principal.times(i), ONE.minus(discount), MONEY_SCALE);
};

export const simulate = (input: MortgageInput): MortgageResult => {
  const i = monthlyRate(input.annualRatePct, input.rateConvention);
  const principal = scale(loanAmountUf(input), MONEY_SCALE);
  const n = termMonths(input);
  const basePayment = payment(principal, i, n);

  const lifeRate = divScale(input.lifeInsuranceMonthlyPct, HUNDRED, RATE_SCALE);
  const fire = scale(input.fireInsuranceMonthlyUf, MONEY_SCALE);

  const byMonth = new Map<number, Prepayment[]>();
  for (const p of input.prepayments) {
    const list = byMonth.get(p.monthNumber) ?? [];
    list.push(p);
    byMonth.set(p.monthNumber, list);
  }

  const rows: AmortizationRow[] = [];
  let balance = principal;
  let currentPayment = basePayment;
  let month = 1;

  // A prepayment can only ever shorten the schedule, so n is a hard cap.
  while (balance.cmp(0) > 0 && month <= n) {
    const opening = balance;
    const interest = scale(opening.times(i), MONEY_SCALE);
    let principalPart = scale(currentPayment.minus(interest), MONEY_SCALE);
    let thisPayment = currentPayment;

    // Final instalment: settle whatever is left rather than overshoot.
    if (principalPart.cmp(opening) >= 0) {
      principalPart = opening;
      thisPayment = scale(principalPart.plus(interest), MONEY_SCALE);
    }
    // A negative-amortising payment would never repay the loan.
    if (principalPart.cmp(0) < 0) principalPart = scale(ZERO, MONEY_SCALE);

    const life = scale(opening.times(lifeRate), MONEY_SCALE);
    let afterPrincipal = scale(opening.minus(principalPart), MONEY_SCALE);

    let prepaid = scale(ZERO, MONEY_SCALE);
    const prepays = byMonth.get(month);
    if (prepays !== undefined && afterPrincipal.cmp(0) > 0) {
      for (const p of prepays) {
        const applied = minOf(scale(p.amountUf, MONEY_SCALE), afterPrincipal);
        if (applied.cmp(0) <= 0) continue;
        prepaid = prepaid.plus(applied);
        afterPrincipal = afterPrincipal.minus(applied);
        if (p.mode === "REDUCE_PAYMENT") {
          const remaining = n - month;
          currentPayment = remaining > 0 ? payment(afterPrincipal, i, remaining) : afterPrincipal;
        }
        // REDUCE_TERM keeps the payment; the loop simply ends sooner.
      }
    }

    rows.push({
      number: month,
      date: plusMonths(input.firstPaymentDate, month - 1),
      openingBalanceUf: opening,
      interestUf: interest,
      principalUf: principalPart,
      paymentUf: thisPayment,
      lifeInsuranceUf: life,
      fireInsuranceUf: fire,
      prepaymentUf: prepaid,
      totalOutflowUf: scale(thisPayment.plus(life).plus(fire).plus(prepaid), MONEY_SCALE),
      closingBalanceUf: afterPrincipal,
    });

    balance = afterPrincipal;
    month++;
  }

  const stampTax = divScale(principal.times(input.stampTaxPct), HUNDRED, MONEY_SCALE);
  const upfront = scale(
    input.originationFeeUf.plus(stampTax).plus(input.otherUpfrontCostsUf),
    MONEY_SCALE,
  );

  const totalInterest = sumScaled(rows.map((r) => r.interestUf), MONEY_SCALE);
  const totalLife = sumScaled(rows.map((r) => r.lifeInsuranceUf), MONEY_SCALE);
  const totalFire = sumScaled(rows.map((r) => r.fireInsuranceUf), MONEY_SCALE);
  const totalPrepay = sumScaled(rows.map((r) => r.prepaymentUf), MONEY_SCALE);
  const totalOutflow = sumScaled(rows.map((r) => r.totalOutflowUf), MONEY_SCALE);

  return {
    input,
    loanAmountUf: principal,
    monthlyRate: i,
    basePaymentUf: basePayment,
    firstTotalPaymentUf: rows[0]?.totalOutflowUf ?? scale(ZERO, MONEY_SCALE),
    schedule: rows,
    totalInterestUf: totalInterest,
    totalLifeInsuranceUf: totalLife,
    totalFireInsuranceUf: totalFire,
    totalPrepaymentsUf: totalPrepay,
    upfrontCostsUf: upfront,
    totalCostUf: scale(totalOutflow.plus(upfront), MONEY_SCALE),
    caePct: solveCae(principal, upfront, rows.map((r) => r.totalOutflowUf)),
    effectiveTermMonths: rows.length,
    monthsSaved: Math.max(0, n - rows.length),
  };
};

/**
 * Carga Anual Equivalente: the annual rate that makes the present value of
 * everything the borrower pays equal the net amount actually received.
 *
 * Solved by bisection on the monthly rate. Returns null when the cash flows
 * have no sign change and therefore no root.
 */
export const solveCae = (
  principal: Money,
  upfrontCosts: Money,
  outflows: readonly Money[],
): Money | null => {
  if (outflows.length === 0 || principal.cmp(0) <= 0) return null;
  const net = Number(principal.minus(upfrontCosts).toString());
  if (net <= 0) return null;
  const flows = outflows.map((f) => Number(f.toString()));

  const npv = (monthly: number): number => {
    let acc = net;
    let discount = 1;
    const step = 1 + monthly;
    for (const f of flows) {
      discount *= step;
      acc -= f / discount;
    }
    return acc;
  };

  let lo = -0.9999;
  let hi = 5;
  if (npv(lo) * npv(hi) > 0) return null;
  for (let k = 0; k < 200; k++) {
    const mid = (lo + hi) / 2;
    if (npv(lo) * npv(mid) <= 0) hi = mid;
    else lo = mid;
  }
  const monthly = (lo + hi) / 2;
  const annual = (Math.pow(1 + monthly, 12) - 1) * 100;
  if (!Number.isFinite(annual)) return null;
  return scale(money(annual.toFixed(10)), 2);
};
